mpu6050/mpu6050.go

```diff
@@ -1,95 +1,251 @@
-package mpu6050 // import "tinygo.org/x/drivers/mpu6050"
+package mpu6050
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"tinygo.org/x/drivers"
+)
 
-import "tinygo.org/x/drivers"
+const DefaultAddress = 0x68
 
-// Device wraps an I2C connection to a MPU6050 device.
-type Device struct {
-	bus     drivers.I2C
-	Address uint16
+type IMUConfig struct {
+	// Use ACCEL_RANGE_2 through ACCEL_RANGE_16.
+	AccRange byte
+	// Use GYRO_RANGE_250 through GYRO_RANGE_2000
+	GyroRange   byte
+	sampleRatio byte // TODO(soypat): expose these as configurable.
+	clkSel      byte
 }
 
-// New creates a new MPU6050 connection. The I2C bus must already be
-// configured.
-//
-// This function only creates the Device object, it does not touch the device.
-func New(bus drivers.I2C) Device {
-	return Device{bus, Address}
+// Dev contains MPU board abstraction for usage
+type Dev struct {
+	conn    drivers.I2C
+	aRange  int32 //Gyroscope FSR acording to SetAccelRange input
+	gRange  int32 //Gyroscope FSR acording to SetGyroRange input
+	data    [14]byte
+	address byte
+}
+
+// New instantiates and initializes a MPU6050 struct without writing/reading
+// i2c bus. Typical I2C MPU6050 address is 0x68.
+func New(bus drivers.I2C, addr uint16) *Dev {
+	p := &Dev{}
+	p.address = uint8(addr)
+	p.conn = bus
+	return p
+}
+
+// Init configures the necessary registers for using the
+// MPU6050. It sets the range of both the accelerometer
+// and the gyroscope, the sample rate, the clock source
+// and wakes up the peripheral.
+func (p *Dev) Init(data IMUConfig) (err error) {
+
+	// setSleep
+	if err = p.SetSleep(false); err != nil {
+		return fmt.Errorf("set sleep: %w", err)
+	}
+	// setClockSource
+	if err = p.SetClockSource(data.clkSel); err != nil {
+		return fmt.Errorf("set clksrc: %w", err)
+	}
+	// setSampleRate
+	if err = p.SetSampleRate(data.sampleRatio); err != nil {
+		return fmt.Errorf("set sample: %w", err)
+	}
+	// setFullScaleGyroRange
+	if err = p.SetGyroRange(data.GyroRange); err != nil {
+		return fmt.Errorf("set gyro: %w", err)
+	}
+	// setFullScaleAccelRange
+	if err = p.SetAccelRange(data.AccRange); err != nil {
+		return fmt.Errorf("set accelrange: %w", err)
+	}
+	return nil
 }
 
-// Connected returns whether a MPU6050 has been found.
-// It does a "who am I" request and checks the response.
-func (d Device) Connected() bool {
-	data := []byte{0}
-	d.bus.ReadRegister(uint8(d.Address), WHO_AM_I, data)
-	return data[0] == 0x68
+// Update fetches the latest data from the MPU6050
+func (p *Dev) Update() (err error) {
+	if err = p.read(ACCEL_XOUT_H, p.data[:]); err != nil {
+		return err
+	}
+	return nil
 }
 
-// Configure sets up the device for communication.
-func (d Device) Configure() error {
-	return d.SetClockSource(CLOCK_INTERNAL)
+// Get treats the MPU6050 like an ADC and returns the raw reading of the channel.
+// Channels 0-2 are acceleration, 3-5 are gyroscope and 6 is the temperature.
+func (d *Dev) RawReading(channel int) int16 {
```

Comment:
Do we need to expose this?
I'd prefer not to expose it, as it is a detail of the specific sensor that is in use. I can't think of any case where you'd want to use the raw value.

Reply:
Actually I have a use case:
I want to use a weak microcontroller to only read sensor's raw data, and send to PC via USB without processing the raw data.

Reply:
Have you measured whether this helps performance? I doubt it will make a significant difference.

Reply:
Not tested yet.

Another use case is that some MPU 6050 chip will enable DMP by default, and when I read acceleration data, it actually returns the computed position data. In that case, maybe those bytes shouldn't be further processed. I don't know whether this is a standard behaviour for MPU 6050.
For MPU 6050 with DMP I also opened an issue #534.

Well, I'm new to embeded programming, sorry if I wrote some stupid questions.

Reply:
I've worked in a similar fashion the last years. Where I can I defer computation to a PC. I could also expose the actual `data` field of Device as `RawData` to get the same effect, probably would be even better for users wanting to squeeze out performance

Reply:
I've applied changes to expose RawData field, which I believe is clearer for users and adds less noise to the API. It also explicitly defines that part of the API as a mpu6050 specific thing since we still do not have generic field APIs in Go, so no risk of locking in an API with this decision.

Reply:
To be very honest, this sounds like premature optimization to me. So until someone gives me the numbers that show it helps in practice, I'd rather _not_ expose these raw values.

If you have a very specific use case, you can of course still read these values manually from the I2C bus.